api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java

```diff
@@ -52,37 +52,44 @@ private ProjectionUtil() {}
     }
   }
 
-  static UnboundPredicate<Integer> truncateIntegerStrict(
-      String name, BoundPredicate<Integer> pred, Transform<Integer, Integer> transform) {
+  static <T> UnboundPredicate<T> truncateIntegerStrict(
+      String name, BoundPredicate<Integer> pred, Transform<Integer, T> transform) {
     int boundary = pred.literal().value();
     switch (pred.op()) {
       case LT:
-        // predicate would be <= the previous partition
-        return predicate(Expression.Operation.LT_EQ, name, transform.apply(boundary) - 1);
+        // Checking if the literal is at the lower partition boundary
+        if (transform.apply(boundary - 1).equals(transform.apply(boundary))) {
+          // Literal is not at lower boundary, for eg: 123, if width is 10
+          return predicate(Expression.Operation.LT, name, transform.apply(boundary - 1));
+        } else {
+          // Literal was at the lower boundary, for eg: 100 if width is 10
+          return predicate(Expression.Operation.LT, name, transform.apply(boundary));
+        }
       case LT_EQ:
-        // Checking if the literal is at the upper partition boundary
         if (transform.apply(boundary + 1).equals(transform.apply(boundary))) {
-          // Literal is not at upper boundary, for eg: 2019-07-02T02:12:34.0000
-          // the predicate can be < 2019-07-01
-          return predicate(Expression.Operation.LT_EQ, name, transform.apply(boundary) - 1);
+          // Literal is not at upper boundary, for eg: 123 if width is 10
+          return predicate(Expression.Operation.LT, name, transform.apply(boundary));
```

Comment:
Same here. I think we can probably simplify some of these cases.

Reply:
Done